File: zig/114.zig
const std = @import("std");

const TreeNode = struct {
    val: i32,
    left: ?*TreeNode,
    right: ?*TreeNode,
};

fn solution(node: ?*TreeNode) ?*TreeNode {
    return flattenTree(node, null);
}

fn flattenTree(node: ?*TreeNode, current_prev: ?*TreeNode) ?*TreeNode {
    if (node == null) {
        return node;
    }

    var prev = current_prev;

    prev = flattenTree(node.?.right, prev);
    prev = flattenTree(node.?.left, prev);

    node.?.right = prev;
    node.?.left = null;

    return node;
}

File: go/114.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func flatten(node *TreeNode) *TreeNode {
	return flattenTree(node, nil)
}

func flattenTree(node *TreeNode, prev *TreeNode) *TreeNode {
	if node == nil {
		return prev
	}

	prev = flattenTree(node.Right, prev)
	prev = flattenTree(node.Left, prev)

	node.Right = prev
	node.Left = nil

	return node
}
